問題のコンポーネントを作成しましたFront/5

x0y14 opened the pull request:
<!--
必ず画面右のメニューからReviewers,Labels,Projectsを設定してください
-->

## 変更点
- 問題画面の親コンポーネントQuestion.vueと、入力されたjsonで切り替わるQestion4taku.vue、QuestionAnaume.vueを作成しました。
Vue2と、Vue3の違いがあまりわかっておらず、Vue2に即した書き方になっておらず、少し汚いです。すみません。
実際に動作可能な状態になっています。大掛かりなバグ探しはしていません。

- 表示テスト用にコンポーネントを表示するComponentsPreviewという画面を作成しました。

- srcの下にtestdataというフォルダを作成しました。backendから送られてくるjsonのサンプルを設置しています。  なお、Question.vueもここからデータを取得しています。

## 対応するissue

<!-- closed #issue番号 のように記載してください -->
closed #5

x0y14 committed:
問題のタイプによって、コンポーネントを切り替える。

x0y14 committed:
少し雑だが、4択問題と、穴埋め問題を動作可能な状態にした。

nanagami1369 commented on front/src/components/QuestionAnaume.vue:
```diff
@@ -0,0 +1,62 @@
+
+      // 参考: https://qiita.com/iwato/items/183e6dd676bf547ea341
+      // 分割対象を残した状態で、分割する。
+      let question_split = text.split(/([])/g);
```
正規表現のSyntax errorが発生してます。
修正をお願いします。
![image](https://user-images.githubusercontent.com/38910015/131437043-8628d982-addf-4779-b73a-f8b2eb4e99bc.png)

x0y14 replied:
スラッシュをつけて対応します！

nanagami1369 reviewed (changes requested):
これ直さないと、
こちらからだとビルドエラーになるので修正をお願いします。

nanagami1369 commented:
あとPRへタイトルをつけるのをお願いします。

x0y14 committed:
問題分割部分の括弧のエラーにスラッシュをつけて対応しました

nanagami1369 commented on front/src/components/Question.vue:
```diff
@@ -0,0 +1,43 @@
+<template>
+    <div class="question-card" id="question-parent">
+        <div v-if="questionDataJson.questionType === '4taku'">
+            <Question4taku v-bind:question="questionDataJson"/>
```
`v-bind:hoge`を使ったものと、その省略記法`:hoge`を使ったものの両方があります。
省略記法に統一をお願いします。

nanagami1369 commented on front/src/components/Question4taku.vue:
```diff
@@ -0,0 +1,32 @@
+      <span></span>
+      <div v-for="v in question.value" :key="v.id">
+        <label class="radio">
+          <input type="radio" name="answer" :value="v" v-model="userChoice">
```
v-bind:hogeを使ったものと、その省略記法:hogeを使ったものの両方があります。
省略記法に統一をお願いします。

`:value`は省略記法ですね

nanagami1369 commented on front/src/views/ComponentsPreviewPage.vue:
```diff
@@ -0,0 +1,34 @@
+  },
+  data () {
+    return {
+      sampleQuestionDataJson: require("@/testdata/question.json")
```
`sampleQuestionDataJson`は、requireで読み込んだ時点で単なるobjectなので変数名からjsonを外して下さい。
この状態だとstring型でパースが必要なのかと思ってしまいます。

nanagami1369 commented on front/src/views/ComponentsPreviewPage.vue:
```diff
@@ -0,0 +1,34 @@
+<template>
+  <div class="components-preview">
+    <div class="question-preview">
+      <QuestionParent v-bind:questionDataJson="sampleQuestionDataJson.q_4taku" />
```
`questionDataJson`は、requireで読み込んだ時点で単なるobjectなので変数名からjsonを外して下さい。
この状態だとstring型の文字列のjsonを渡すのかと思ってしまいます。

nanagami1369 commented on front/src/components/QuestionAnaume.vue:
```diff
@@ -0,0 +1,62 @@
+<template>
+    <div id="question-anaume">
+      <h3>title : {{question.name}}</h3>
+      <div v-for="wordData in parseQuestion(question.question)" :key="wordData.id">
```
## `parseQuestion`について
`methods`は主にclick等のイベント発生時に使うものです。
こういった表示に使う場合は`computed`を使って下さい

nanagami1369 replied:
`wordData.id`は単なる文字列配列ですので
v.id == undefinedですね

nanagami1369 commented on front/src/components/Question4taku.vue:
```diff
@@ -0,0 +1,32 @@
+      <div v-for="v in question.value" :key="v.id">
+        <label class="radio">
+          <input type="radio" name="answer" :value="v" v-model="userChoice">
+          {{ v }}
```
### `v`について

一文字変数はあまりつかわないで欲しい。
意味が類推しづらい
この場合は`value`、もしくは`val`で

for文の`i`とかなら確実に`index`だとわかるみたいな例外はあるけど基本無しで

nanagami1369 commented on front/src/components/QuestionAnaume.vue:
```diff
@@ -0,0 +1,62 @@
+      let holeIndex = 0;
+      for (let i=0; i<question_split.length; i++) {
+        let word = question_split[i];
+        let d = {
```
`v`と同じ理由で修正お願いします。

nanagami1369 reviewed (changes requested):
すみません。修正お願いします。
それから修正が終わったら、developをmargeするコミットをして下さい。
後、さっきマージしてもらったPRのコミットにformatterの設定も入ってますので
formatツイートもお願いします。

x0y14 committed:
- v-bind:~を使用した形から、:~を使用した省略形に統一しました。
- 穴埋め問題の表示にmethodを使うのではなくcomputedを使うようにしました。
- １文字変数だったところをわかりやすく(v->value)
- 変数名に余分についていた、関係ない型を削除しました(DataJson->Data)

nanagami1369 commented on front/src/components/Question4taku.vue:
```diff
@@ -0,0 +1,32 @@
+      <h3 class="title" > {{ question.name }} </h3>
+      <h3>Q. {{ question.question }}</h3>
+      <span></span>
+      <div v-for="v in question.value" :key="v.id">
```
`question.value`は単なる文字列配列ですので
v.id == undefinedですね

x0y14 replied:
では、v-forの部分を(value, index)に変更し、keyをindexに変更します！

nanagami1369 replied:
他にもあるので修正お願いします

x0y14 replied:
同じ理由で、なんか箇所か修正を行いました！

x0y14 committed:
v-forのkeyに不正な値が入る可能性があったので、indexを代入するようにしました

nanagami1369 reviewed (changes requested):
気がついた修正点は以上になります。

後は、最新のdevelopブランチをマージしてコンフリクトの解決と、
format commitをお願いします。

x0y14 committed:
- 1文字変数の名前をわかりやすいものにしました。
- スネークケースと、キャメルケースが混在していたため、キャメルケースに統一しました

x0y14 committed:
developをこちらにマージし、コンフリクトを解決しました。

x0y14 committed:
整形が行われていなかったので、整形しました

nanagami1369 reviewed (changes requested):
実行時エラーの修正をお願いします。

![image](https://user-images.githubusercontent.com/38910015/131446054-1dd777a2-cd5c-485e-954f-85f01d1418a6.png)

x0y14 committed:
親要素が同じv-forで発生するkeyエラーに対応しました
参考: https://qiita.com/ysKuga/items/fa2ba12b10bade86da36

nanagami1369 reviewed (approved):
OKです。お疲れ様でした。
